File: 2016/1B/1/digits.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include <fstream>
#include <sstream>
using std::string;
using std::vector;
using std::ifstream;
using std::stringstream;
using std::cout;
using std::endl;
using std::size_t;
using std::to_string;
using std::rotate;
using std::stoi;

typedef vector<vector<string> > VecOfVecOfStrings;

vector<string> V = {"ZERO", "ONE", "TWO",   "THREE", "FOUR",
                    "FIVE", "SIX", "SEVEN", "EIGHT", "NINE"
                   };

struct example {
    string number;
    string tel;
    int vec_shift;
};

VecOfVecOfStrings read_file(string file_name, bool print_console = false) {
    string line;
    VecOfVecOfStrings v;
    ifstream myfile(file_name);
    if (myfile.is_open()) {
        while (getline(myfile, line)) {
            string buf;
            stringstream ss(line);
            vector<string> tokens;
            while (ss >> buf) tokens.push_back(buf);
            v.push_back(tokens);
        }
        myfile.close();
    } else
        cout << "Unable to open file" << endl;
    if (print_console) {
        for (const vector<string>& vec : v) {
            for (string x : vec) cout << x << ' ';
            cout << endl;
        }
    }
    return v;
}

example digit_in_string(string digit, example exmpl) {
    if (exmpl.number.empty()) return exmpl;
    example new_exmpl = exmpl;
    size_t found;
    for (char& c : digit) {
        found = new_exmpl.number.find(c);
        if (found == string::npos) return exmpl;
        new_exmpl.number.erase(new_exmpl.number.begin() + found);
    }
    vector<string>::iterator it;
    it = find(V.begin(), V.end(), digit);
    size_t i = it - V.begin();
    new_exmpl.tel += to_string(i);
    return digit_in_string(digit, new_exmpl);
}

void test_all(example& exmpl) {
    string temp_str = exmpl.number;
    vector<string> new_V = V;
    rotate(new_V.begin(), new_V.begin() + exmpl.vec_shift, new_V.end());
    for (int i = 0; i < new_V.size(); ++i) {
        exmpl = digit_in_string(new_V[i], exmpl);
    }
    if (exmpl.number != "") {
        exmpl.vec_shift += 1;
        exmpl.number = temp_str;
        exmpl.tel = "";
        test_all(exmpl);
    } else {
        return;
    }
}

int main(int argc, char const* argv[]) {
    VecOfVecOfStrings f = read_file("A2.in");
    // for (int i = 0; i < stoi(f[0][0]); ++i) {
    example test;
    int i = 98;
    test.number = f[i + 1][0];
    test.vec_shift = 0;
    test.tel = "";
    test_all(test);
    cout << test.tel << endl;
    // }
    return 0;
}

File: 2016/1B/1/digits.py
from __future__ import division
from collections import Counter
import itertools
import math


def read_file(file_name):
    with open(file_name, 'r') as f:
        lines = f.read().splitlines()
    return lines


# def multiple_output(n, fun):
#     for i in xrange(n):
#         write_output(i + 1, fun(lines[i + 1]))

dictionary = ["ZERO", "ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN", "EIGHT", "NINE"]


def if_digit_in_string(digit, string, i):
    # if string is '':
        # return i, ''
    newstr = string
    for ch in digit:
        if ch not in newstr:
            return i, string
        newstr = update_string(ch, newstr)
    i+=1
    # if newstr:
    i, newstr = if_digit_in_string(digit, newstr, i)
    # else:
    return i, newstr



def update_string(digit, string):
    newstr = string
    for ch in digit:
        newstr = newstr.replace(ch, "", 1)
    return newstr


def find_numbers(string):
    number = ''
    newstr = string
    for j, d in enumerate(dictionary):
        i, newstr = if_digit_in_string(d, newstr,0)
        number += i*str(j)
    if newstr:
        newstr = string
        number = ''
        for j, d in enumerate(reversed(dictionary)):
            i, newstr = if_digit_in_string(d, newstr,0)
            number += i*str(9-j)
    return ''.join(sorted(number))


if __name__ == '__main__':
    lines = read_file("B1.in")
    for n in xrange(0, int(lines[0])):
    # n = 98
        string = lines[n + 1]
        print("Case #%d: %s" % (n + 1, find_numbers(string)))

